Add support for `markdown` format output

sandrobonazzola opened the pull request:
Add markdown format output and enable it for bugzilla, github and gitlab.

kwk commented on did/cli.py:
```diff
@@ -149,8 +149,15 @@ def parse(self):
             period, opt.since, opt.until.date - delta(days=1))
+        if opt.format == "md":
+            # in markdown first line must be an header
+            header = "# " + header
```
I'd suggest to use the alternate syntax for headings as described here: https://www.markdownguide.org/basic-syntax/#alternate-syntax. Then you can use `did`'s output in commit messages as well. The `#` headers are always consumed as comments which is a bit unfortunate.

sandrobonazzola replied:
done

kwk commented on did/utils.py:
```diff
@@ -185,7 +185,7 @@ def item(text, level=0, options=None):
         return
     # Four space for each level, additional space for wiki format
-    indent = level * 4
+    indent = level * (4 if options.format != "md" else 2)
```
If I'm interpreting this correctly you want to have `2` spaces for the markdown format. I suggest to drop the syntactic sugar and just make it more obvious what's the default and what' special for markdown.

```suggestion
    indent = level * 4
    if options.format == "md":
        indent = level * 2
```

kwk reviewed (commented):
Niced idea. I have some just some minor thoughts.

psss commented on README.rst:
```diff
@@ -86,12 +86,13 @@ The default output is plain text of maximum width 79 characters.
 shortening altogether use ``--width=0``. The default width value
 can be saved in the config file as well. Use ``--format=wiki`` to
-enable simple MoinMoin wiki syntax. For stats which support them,
+enable simple MoinMoin wiki syntax or ``--format=wiki`` to enable
```
```suggestion
enable simple MoinMoin wiki syntax or ``--format=md`` to enable
```

psss replied:
Or, perhaps, would you mind if we use `--format=markdown` instead of the abbreviation?

sandrobonazzola replied:
ok, I'll send a patch

psss commented on README.rst:
```diff
@@ -86,12 +86,13 @@ The default output is plain text of maximum width 79 characters.
 
---format=FMT
-    Output style, possible values: text (default) or wiki
+--format {text,md,wiki}
```
```suggestion
--format {text|md|wiki}
```

Perhaps, `|` is more commonly used to separate alternative values?

sandrobonazzola replied:
This is the separator used by `argparse` in the `--help` output. I've no problem with your suggestion for the README if you're ok it won't match the help output. Let me know what you prefer.

psss commented on did/plugins/gitlab.py:
```diff
@@ -176,9 +177,18 @@ def iid(self):
+                self.gitlabapi.url,
+                self.project['path_with_namespace'],
+                ("issues"
+                    if self.data['target_type'] == 'Issue'
```
I tried this with my account and for this issue:

* https://gitlab.com/testing-farm/general/-/issues/41

I got a link to a (non-existent) merge request instead:

* https://gitlab.com/testing-farm/general/-/merge_requests/41

Quickly checked the `self.data` and I see this:

    'noteable_type': 'Issue'
    'target_type': 'Note'

Perhaps, we should use `noteable_type` here instead?

sandrobonazzola replied:
Sorry, missed a case, fixing.

psss commented on did/plugins/gitlab.py:
```diff
@@ -223,7 +233,7 @@ def fetch(self):
             'Issue', 'opened')
         self.stats = [
-            Issue(issue, self.parent.gitlab)
+            Issue(issue, self.parent.gitlab, self.parent)
```
So if we hand over the whole `parent` now, we could extract the `self.parent.gitlab` in the constructor as well?

sandrobonazzola replied:
yes, I'll change that.

psss commented on did/cli.py:
```diff
@@ -149,8 +149,17 @@ def parse(self):
+        header = "Status report for {0} ({1} to {2})".format(
             period, opt.since, opt.until.date - delta(days=1))
+        if opt.format == "md":
+            # in markdown first line must be an header
```
```suggestion
            # In markdown the first line must be a header
```

psss commented on did/cli.py:
```diff
@@ -149,8 +149,17 @@ def parse(self):
+            # output in commit messages as well
+            header = f"{header}\n{'=' * len(header)}"
+        else:
+            # in markdown no trailing punctuation is allowed in headings
```
```suggestion
            # In markdown no trailing punctuation is allowed in headings
```

Just consistent comments...

psss reviewed (changes requested):
Nice improvement! Thanks for implementing this. Added just a few, rather minor, comments.

psss commented on README.rst:
```diff
@@ -86,12 +86,13 @@ The default output is plain text of maximum width 79 characters.
 shortening altogether use ``--width=0``. The default width value
 can be saved in the config file as well. Use ``--format=wiki`` to
-enable simple MoinMoin wiki syntax. For stats which support them,
+enable simple MoinMoin wiki syntax or ``--format=wiki`` to enable
```
```suggestion
enable simple MoinMoin wiki syntax or ``--format=markdown`` to enable
```

Last remnant.

psss commented:
One more thing: It seems that GitHub pull requests are created with url leading to issues:

    did --format markdown --width 0 --github-pull-requests-created

The output looks like this:

Status report for the week 36 (2023-09-04 to 2023-09-10)
========================================================

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 Petr Šplíchal <psplicha@redhat.com>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Pull requests created on github: 5
  * [psss/did#320](https://github.com/psss/did/issues/320) - Move line length recommendation to proper section
  * [teemtee/tmt#2315](https://github.com/teemtee/tmt/issues/2315) - Make the upload to PyPI working again
  * [teemtee/tmt#2314](https://github.com/teemtee/tmt/issues/2314) - Adjust test coverage for deep beakerlib libraries
  * [teemtee/tmt#2312](https://github.com/teemtee/tmt/issues/2312) - Release tmt-1.27.0
  * [teemtee/tmt#2308](https://github.com/teemtee/tmt/issues/2308) - Disable `systemd-resolved` to prevent dns failures

GitHub automatically redirects these to the proper url, so it's not that important to get fixed, but probably would be a bit nicer/cleaner. @sandrobonazzola, what do you think?

sandrobonazzola commented:
> GitHub automatically redirects these to the proper url, so it's not that important to get fixed, but probably would be a bit nicer/cleaner. @sandrobonazzola, what do you think?

I've kind of cheated there to keep the code simpler but yes, some more parsing can be done to get avoid the redirect. I would consider doing it in another PR as enhancement when I'll have some more time to dedicate to this.

sandrobonazzola commented:
Nevermind, found a more useful way to do it, pushing a new commit

psss commented:
> Nevermind, found a more useful way to do it, pushing a new commit

Perfect, thanks!

psss reviewed (approved):
Thanks for addressing all comments, added just a tiny adjustment in 2c505c0.

sandrobonazzola commented:
> Thanks for addressing all comments, added just a tiny adjustment in [2c505c0](https://github.com/psss/did/commit/2c505c0600fbf3adcdec1e207ed9f45ee462f6b1).

LGTM, thanks

psss commented:
For some reason, the `--format` option is incorrectly render in the docs:

![image](https://github.com/psss/did/assets/2803150/c505394e-1f71-4900-a486-0e51058d18cb)

Turning off `smartquotes` in 273d33c should fix it.
